std/file.d

```diff
@@ -3200,6 +3200,219 @@ version(Posix) unittest //issue 11434
     assert(getAttributes(t2) == octal!100767);
 }
 
+/***************************************************
+Copy file or directory $(D from) to $(D to). When a directory is copied, its
+contents are copied as well, recursively.
+
+File timestamps are preserved, except for symbolic links.
+File attributes are preserved, if $(D preserve) equals $(D PreserveAttributes.yes).
+On Windows only $(D PreserveAttributes.yes) (the default on Windows) is supported.
+
+Existing target files are overwritten. Existing target directories are written
+into.
+
+With symbolic links, the links themselves are copied. That is, for every
+symbolic link in $(D from), a new link with the same target path is created in
+$(D to). Relative target paths are not changed. They may be invalidated by the
+copying, when they point outside of the copied directory tree.
+
+Special files (e.g. device files, FIFOs) cannot be copied with $(D copyRecurse).
+
+Params:
+    from = string or range of characters representing the existing
+        file/directory name
+    to = string or range of characters representing the target file/directory
+        name
+    preserve = whether to preserve the file attributes
+
+Throws: $(D FileException) on error.
+ */
+void copyRecurse(string from, string to,
+    PreserveAttributes preserve = preserveAttributesDefault)
+{
+    // Remove any trailing directory separators.
+    from = buildNormalizedPath(from);
+    to = buildNormalizedPath(to);
+
+    immutable attrs = getLinkAttributes(from);
+
+    if (attrIsFile(attrs)) copy(from, to, preserve);
+    else if (attrIsDir(attrs))
+    {
+        mkdirRecurse(std.conv.to!string(to));
+        
+        if (preserve == PreserveAttributes.yes) setAttributes(to, attrs);
+        
+        foreach (entry; dirEntries(from, SpanMode.shallow, false))
+        {
+            immutable path = entry.name[from.length + 1 .. $];
+                // + 1 for the directory separator
+            copyRecurse(entry.name, buildPath(to, path), preserve);
+        }
+        
+        SysTime accessTime, modificationTime;
+        getTimes(from, accessTime, modificationTime);
+        setTimes(to, accessTime, modificationTime);
+    }
+    else if (attrIsSymlink(attrs))
+    {
+        version (Posix)
+        {
+            if (exists(to)) remove(to);
+            symlink(readLink(from), to);
+            if (preserve == PreserveAttributes.yes)
+            {
+                /*NOTE: Would copy attributes of the symlink here, but there
+                is no setLinkAttributes here, and no fchmodat in the C headers.
+                Just enforcing that the new attributes are the same as the old
+                ones, for now. */
```

Comment:
I'm not sure we should settle on any user-visible behavior "for now", because then fixing it becomes a breaking change.